File: src/main/java/cz/encircled/eplayer/core/ApplicationCore.java
package cz.encircled.eplayer.core;

import cz.encircled.eplayer.common.Constants;
import cz.encircled.eplayer.model.MediaType;
import cz.encircled.eplayer.service.*;
import cz.encircled.eplayer.service.event.EventObserver;
import cz.encircled.eplayer.service.event.EventObserverImpl;
import cz.encircled.eplayer.service.gui.FxViewService;
import cz.encircled.eplayer.service.gui.ViewService;
import cz.encircled.eplayer.util.IOUtil;
import cz.encircled.eplayer.view.fx.FxView;
import javafx.application.Platform;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.annotations.NotNull;
import uk.co.caprica.vlcj.player.MediaPlayer;

import java.io.File;

public class ApplicationCore {

    public static final String APP_DOCUMENTS_ROOT = System.getenv("APPDATA") + "\\EPlayer\\";

    public static final String SCREENS_FOLDER = APP_DOCUMENTS_ROOT + "\\frames\\";

    public static final String URL_FILE_PREFIX = "file:\\\\\\";

    private static final Logger log = LogManager.getLogger();

    private MediaService mediaService;

    private CacheService cacheService;

    private ViewService viewService;

    private EventObserver eventObserver;

    private SeriesFinder seriesFinder;

    private FolderScanService folderScanService;

    private FxView fxView;

    public ApplicationCore() {
        cacheService = new JsonCacheService();
        folderScanService = new OnDemandFolderScanner(this);
        eventObserver = new EventObserverImpl();
        IOUtil.createIfMissing(APP_DOCUMENTS_ROOT, true, false);
        IOUtil.createIfMissing(SCREENS_FOLDER, true, false);
        addCloseHook();
    }

    @NotNull
    public static String getScreenshotLocation(@NotNull MediaType mediaType) {
        return SCREENS_FOLDER + mediaType.getName() + mediaType.getSize() + ".png";
    }

    @NotNull
    public static String getScreenshotURL(@NotNull MediaType mediaType) {
        String fileLocation = getScreenshotLocation(mediaType);
        return new File(fileLocation).exists() ? URL_FILE_PREFIX + fileLocation : "";
    }

    public void initFx(MediaPlayer mediaPlayer) {
        mediaService = new VLCMediaService(this, mediaPlayer);
    }

    public void init(FxView fxView) {
        this.fxView = fxView;
        cacheService.init();
        seriesFinder = new SeriesFinder();
        viewService = new FxViewService(fxView);
    }

    public void openQuickNavi() {
        new Thread(() -> {
            mediaService.pause();
            viewService.switchToQuickNavi();
            mediaService.updateCurrentMediaInCache();
            mediaService.stop();
            cacheService.save();
        }).start();
    }

    public void back() {
        // TODO
        if (fxView.isPlayerScene()) {
            if (fxView.isFullScreen()) {
//                fxView.setFullScreen(false);
                new Thread(mediaService::updateCurrentMediaInCache).start();
            } else
                openQuickNavi();
        }
    }

    public void playLast() {
        MediaType media = cacheService.getLastByWatchDate();
        if (media != null) {
            mediaService.play(media);
        }
    }

    public void exit() {
        Platform.exit();
        System.exit(Constants.ZERO);
    }

    public SeriesFinder getSeriesFinder() {
        return seriesFinder;
    }

    public MediaService getMediaService() {
        return mediaService;
    }

    public CacheService getCacheService() {
        return cacheService;
    }

    public ViewService getViewService() {
        return viewService;
    }

    public FolderScanService getFolderScanService() {
        return folderScanService;
    }

    public EventObserver getEventObserver() {
        return eventObserver;
    }

    private void addCloseHook() {
        Runtime.getRuntime().addShutdownHook(new Thread() {
            @Override
            public void run() {
                log.debug("Close hook start");
                mediaService.updateCurrentMediaInCache();
                cacheService.save();
                mediaService.releasePlayer();
                log.debug("Close hook finished");
            }
        });
        log.trace("Close hook added");
    }

}

File: src/main/java/cz/encircled/eplayer/core/ApplicationCore.kt
package cz.encircled.eplayer.core

import cz.encircled.eplayer.common.Constants
import cz.encircled.eplayer.model.AppSettings
import cz.encircled.eplayer.model.MediaFile
import cz.encircled.eplayer.remote.RemoteControlHandler
import cz.encircled.eplayer.remote.RemoteControlHttpServer
import cz.encircled.eplayer.remote.RemoteControlHttpServerImpl
import cz.encircled.eplayer.service.*
import cz.encircled.eplayer.service.event.Event
import cz.encircled.eplayer.util.IOUtil
import cz.encircled.eplayer.util.LocalizationProvider
import cz.encircled.eplayer.util.TimeMeasure.measure
import cz.encircled.eplayer.view.AppView
import cz.encircled.eplayer.view.swing.AppActions
import javafx.application.Platform
import org.apache.logging.log4j.LogManager
import kotlin.system.exitProcess

/**
 * TODO list
 *
 * Series:
 * - Auto next episode
 * title changes on pause
 *
 * Check display changes
 * Scan all folders on start
 *
 * show info messages: name, subtitle, audio track etc
 *
 * CHECK TRUEHD DOLBY
 */
class ApplicationCore(
    private val remoteControlServerCreator: (RemoteControlHandler) -> RemoteControlHttpServer = {
        RemoteControlHttpServerImpl(it)
    }
) {

    val settings: AppSettings

    val ioUtil: IOUtil = IOUtil()

    lateinit var cacheService: CacheService

    lateinit var folderScanService: FolderScanService

    lateinit var mediaService: MediaService

    lateinit var appView: AppView

    lateinit var appActions: AppActions

    lateinit var seriesFinder: SeriesFinder

    lateinit var metaInfoService: MetadataInfoService

    lateinit var mediaSettingsSuggestions: MediaSettingsSuggestions

    private lateinit var remoteControlServer: RemoteControlHttpServer

    init {
        ioUtil.createIfMissing(APP_DOCUMENTS_ROOT, true)
        ioUtil.createIfMissing(SCREENS_FOLDER, true)

        settings = try {
            ioUtil.getSettings()
        } catch (e: Exception) {
            throw IllegalStateException(e)
        }
        settings.core = this
        LocalizationProvider.init(settings)
    }

    fun delayedInit(appView: AppView) {
        measure("AppCore init") {
            this.appView = appView
            this.metaInfoService = JavacvMetadataInfoService()
            this.mediaSettingsSuggestions = MediaSettingsSuggestionsImpl()
            this.cacheService = JsonCacheService(this)
            this.folderScanService = OnDemandFolderScanner(this)
            this.seriesFinder = SeriesFinder()
            this.appActions = appView.actions
            this.remoteControlServer = remoteControlServerCreator.invoke(appActions)
            this.mediaService = VLCMediaService(this)

            Event.audioPassThroughChange.listen {
                resetPlayer(appView)
            }
        }

        measure("AppCore init player") {
            resetPlayer(appView)
            addCloseHook()
        }

        Event.contextInitialized.fire(this)
    }

    fun exit() {
        Platform.exit()
        exitProcess(Constants.ZERO)
    }

    private fun resetPlayer(appView: AppView) {
        val currentMedia = mediaService.currentMedia()
        appView.setMediaPlayer(mediaService.createPlayer())
        if (currentMedia != null) mediaService.play(currentMedia)
    }

    private fun addCloseHook() {
        Runtime.getRuntime().addShutdownHook(object : Thread() {
            override fun run() {
                stopApp()
            }
        })
        log.trace("Close hook added")
    }

    fun stopApp() {
        log.debug("Close hook start")
        mediaService.stop()
        cacheService.save()
        mediaService.releasePlayer()
        remoteControlServer.stop()
        log.debug("Close hook finished")
    }

    companion object {
        val APP_DOCUMENTS_ROOT = System.getenv("APPDATA") + "\\EPlayer\\"

        val SCREENS_FOLDER = "$APP_DOCUMENTS_ROOT\\frames\\"

        const val URL_FILE_PREFIX = "file:\\\\\\"

        private val log = LogManager.getLogger()

        fun getScreenshotLocation(mediaFile: MediaFile): String {
            return SCREENS_FOLDER + mediaFile.name + mediaFile.size + ".png"
        }

        @JvmStatic
        fun getScreenshotURL(mediaFile: MediaFile): String {
            val fileLocation = getScreenshotLocation(mediaFile)
            return URL_FILE_PREFIX + fileLocation
        }
    }

}